[Bug]Fix: Cannot display main screen on web

doannimble opened the pull request:
## What happened 👀

Cannot display the main screen on web, due to the Platform operator does not support on web when calling (Platform.isAndroid and Platform.isIOS)

## Insight 📝

- Create `PlatformUtils` to check if the current platform`kIsWeb` first, then check if it is mobile later
- Update `main screen` using `PlatformUtils` to display number of items
- Update `main_header` to display `Setup Display` button (that support on mobile only so far)

## Proof Of Work 📹

<img width="800" alt="Screenshot 2023-08-04 at 4 42 27 PM" src="https://github.com/nimblehq/showcase-flutter-build-web-desktop/assets/88996251/0c2763eb-eb45-45ba-afe4-843cc796b545">

Wadeewee commented on lib/util/platform_utils.dart:
```diff
@@ -0,0 +1,12 @@
+
+class PlatformUtils {
+  static bool get isMobile {
+    if (kIsWeb) {
+      return false;
+    } else {
+      return Platform.isIOS || Platform.isAndroid;
+    }
```
We also have window platform, Can we update something like this 🙏 

```suggestion
    if (Platform.isIOS || Platform.isAndroid) {
      return true;
    } else {
      return false;
    }
```

doannimble replied:
@Wadeewee If it is `mac` or `window` platform, the check `Platform.isIOS || Platform.isAndroid` will return `false` 😄

Wadeewee replied:
@doannimble, It meets our expectations because we are using the same UI For building `web`, `macOs`, and `window desktop` platforms 🙏

doannimble replied:
@Wadeewee This function only check it is mobile or not, so I think we can keep the current approach. It meets all our expectations too 🙏

Wadeewee replied:
> @Wadeewee This function only check it is mobile or not, so I think we can keep the current approach. It meets all our expectations too 

@doannimble May I know why we need to update the UI and behavior on the `Web` and `Windows` platform 🤔 

- Here is the original version from the main branch when running on the `web` platform:

https://github.com/nimblehq/showcase-flutter-build-web-desktop/assets/28002315/b0a3c51a-d7b8-4922-b52b-f3bf8ad83963

- And here is the version from the current branch of this PR:

https://github.com/nimblehq/showcase-flutter-build-web-desktop/assets/28002315/54971177-3a8e-40d8-a970-d991238fbe16

Furthermore, The UI on the main screen was broken when resizing the screen, and here is the error:

```
======== Exception caught by rendering library =====================================================
The following assertion was thrown during layout:
A RenderFlex overflowed by 17 pixels on the bottom.

The relevant error-causing widget was:
  Column Column:file:///Users/pooh/StudioProjects/showcase-flutter-build-web-desktop/lib/screen/main/main_screen.dart:123:16
The overflowing RenderFlex has an orientation of Axis.vertical.
The edge of the RenderFlex that is overflowing has been marked in the rendering with a yellow and black striped pattern. This is usually caused by the contents being too big for the RenderFlex.

Consider applying a flex factor (e.g. using an Expanded widget) to force the children of the RenderFlex to fit within the available space instead of being sized to their natural size.
This is considered an error condition because it indicates that there is content that cannot be seen. If the content is legitimately bigger than the available space, consider clipping it with a ClipRect widget before putting it in the flex, or using a scrollable container rather than a Flex, like a ListView.

The specific RenderFlex in question is: RenderFlex#29307 relayoutBoundary=up8 OVERFLOWING
...  needs compositing
...  parentData: <none> (can use size)
...  constraints: BoxConstraints(0.0<=w<=732.0, h=105.0)
...  size: Size(732.0, 105.0)
...  direction: vertical
...  mainAxisAlignment: start
...  mainAxisSize: max
...  crossAxisAlignment: start
...  textDirection: ltr
...  verticalDirection: down
◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤
====================================================================================================
```

It seems the current version changed the design and behavior of  the `web` and `Windows` platforms and It's not a requirement of them 🙈

doannimble replied:
@Wadeewee It is the issue of code merging. The behavior on mobile is different from web/desktop, I will create another PR to fix it.
In this PR, we will fix the main display first 🙏

doannimble replied:
@Wadeewee FYI: Fixed at https://github.com/nimblehq/showcase-flutter-build-web-desktop/pull/59

Wadeewee reviewed (commented):
Just a small concern for the rest LGTM 🚀

hoangnguyen92dn commented on lib/util/platform_utils.dart:
```diff
@@ -0,0 +1,12 @@
+
+class PlatformUtils {
+  static bool get isMobile {
+    if (kIsWeb) {
+      return false;
+    } else {
+      return Platform.isIOS || Platform.isAndroid;
+    }
```
Agree 🤣

doannimble replied:
@luongvo When running on web, calling `Platform.isIOS || Platform.isAndroid` will cause the issue as we got. I added the reason at insight 😄 
cc @hoangnguyen92dn

luongvo replied:
@doannimble So how about this? It's too simple.

```suggestion
    return !kIsWeb && (Platform.isIOS || Platform.isAndroid);
```

doannimble replied:
@luongvo It is a shortened code. However, hard to readability imo 🤔 . Btw, I'm okay with both, if our team loves this style, I will follow.
cc @hoangnguyen92dn @Wadeewee @kaungkhantsoe @ducbm051291 @markgravity

doannimble commented on lib/screen/main/main_screen.dart:
```diff
@@ -427,6 +427,7 @@ class _MainScreenState extends ConsumerState<MainScreen> {
   }
 
   void setUpSecondaryDisplay() async {
+    if (!Platform.isIOS || !Platform.isAndroid) return;
```
@luongvo Ah yes, I've forgotten it 🙏

luongvo commented:
@doannimble I think you should exchange the content of `What happened` and `Insight` sections 🙏

doannimble commented:
> @doannimble I think you should exchange the content of `What happened` and `Insight` sections 🙏

@luongvo For `What happned`, I added the general content (root caused and way to fix). With `Insight section, I added the detail.
I'm not sure what is wrong here 🤔 
https://nimblehq.co/compass/development/code-reviews/author-guidelines/

luongvo commented:
> > @doannimble I think you should exchange the content of `What happened` and `Insight` sections 🙏
> 
> @luongvo For `What happned`, I added the general content (root caused and way to fix). With `Insight` section, I added the detail. I'm not sure what I'm doing wrong here. Could you please point me out? 🙏 [nimblehq.co/compass/development/code-reviews/author-guidelines](https://nimblehq.co/compass/development/code-reviews/author-guidelines/)

It's quite clear from the Compass's documentation, I think:
- What happened: what we did, which changes
  - Create `PlatformUtils` to check if the current platform`kIsWeb` first, then check if it is mobile later
  - Update `main screen` using `PlatformUtils` to display number of items
  - Update `main_header` to display `Setup Display` button (that support on mobile only so far)
- Insight: details, additional info for what happened
  - Cannot display the main screen on web, due to the Platform operator does not support on web when calling (Platform.isAndroid and Platform.isIOS). Hence, we need to implement another platform check.

doannimble commented:
> > > @doannimble I think you should exchange the content of `What happened` and `Insight` sections 🙏
> > 
> > 
> > @luongvo For `What happned`, I added the general content (root caused and way to fix). With `Insight` section, I added the detail. I'm not sure what I'm doing wrong here. Could you please point me out? 🙏 [nimblehq.co/compass/development/code-reviews/author-guidelines](https://nimblehq.co/compass/development/code-reviews/author-guidelines/)
> 
> It's quite clear from the Compass's documentation, I think:
> 
> * What happened: what we did, which changes
>   
>   * Create `PlatformUtils` to check if the current platform`kIsWeb` first, then check if it is mobile later
>   * Update `main screen` using `PlatformUtils` to display number of items
>   * Update `main_header` to display `Setup Display` button (that support on mobile only so far)
> * Insight: details, additional info for what happened
>   
>   * Cannot display the main screen on web, due to the Platform operator does not support on web when calling (Platform.isAndroid and Platform.isIOS). Hence, we need to implement another platform check.

@luongvo Hmm, I'm getting a confusing about this. According to the document:

luongvo commented on lib/screen/main/main_screen.dart:
```diff
@@ -427,6 +425,7 @@ class _MainScreenState extends ConsumerState<MainScreen> {
   }
 
   void setUpSecondaryDisplay() async {
+    if (!PlatformUtils.isMobile) return;
```
We already have a check to build the `setUpDisplay` on the header only on mobile. Do we really need to check 1 more time when handling its callback?

doannimble replied:
@luongvo Just to make sure no one is able to call it accidentally. But agree with you, not really need 👌

luongvo commented on lib/screen/main/main_header.dart:
```diff
@@ -3,9 +3,11 @@ import 'package:showcase_flutter_build_web_desktop/extension/context_extension.d
+import 'package:showcase_flutter_build_web_desktop/util/platform_utils.dart';
 
 class MainHeader extends StatelessWidget {
   final VoidCallback setUpDisplay;
```
Optional: As it's an callback event, should it be `setUpDisplayClicked` or `setUpDisplayPressed`?
